### PR TITLE
Replace some `CrateStore` trait methods with hooks. 

### PR DESCRIPTION
**Pull request by oli-obk**
Just like with the `CrateStore` trait, this avoids the cyclic definition issues with `CStore` being
defined after TyCtxt, but needing to be used in TyCtxt.

### CODE REVIEW

File: compiler/rustc_middle/src/query/on_disk_cache.rs [hunk 1]
```diff
@@ -492,9 +492,7 @@ impl<'a, 'tcx> CacheDecoder<'a, 'tcx> {
                 // expansion, so we use `import_source_files` to ensure that the foreign
                 // source files are actually imported before we call `source_file_by_stable_id`.
                 if source_file_cnum != LOCAL_CRATE {
-                    self.tcx
-                        .cstore_untracked()
-                        .import_source_files(self.tcx.sess, source_file_cnum);
+                    let () = self.tcx.import_source_files(source_file_cnum);
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/query/on_disk_cache.rs:495]:
```suggestion
                    self.tcx.import_source_files(source_file_cnum);
```
Not sure why the `()`.

**Review by petrochenkov — APPROVED** [review]

File: compiler/rustc_middle/src/query/on_disk_cache.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/query/on_disk_cache.rs:495]: 🙈 because of the `must_use` that is set on all hooks

**Comment by RalfJung** [comment]
> This is work towards unblocking https://github.com/rust-lang/rust/pull/122993 (the next step is giving Span deserialization access to the TyCtxt)

FWIW I have closed that PR as there's an alternative, simpler approach; if this is a lot of work it may not be worth it.

**Comment by oli-obk** [comment]
This is work that I wanted to do anyway. It will eventually allow cleanups like getting rid of the `Option<&Session>` in deserialization and just using `TyCtxt` everywhere

**Comment by oli-obk** [comment]
@rustbot label: +perf-regression-triaged

https://github.com/rust-lang/rust/pull/123099#issuecomment-2043604734

**Pull request merged**